src/libstd/dynamic_lib.rs

```diff
@@ -283,18 +288,56 @@ pub mod dl {
     use string::String;
     use vec::Vec;
 
-    pub unsafe fn open_external(filename: &[u8]) -> *mut u8 {
-        // Windows expects Unicode data
-        let filename_str = str::from_utf8(filename).unwrap();
-        let mut filename_str: Vec<u16> = filename_str.utf16_units().collect();
-        filename_str.push(0);
-        LoadLibraryW(filename_str.as_ptr() as *const libc::c_void) as *mut u8
-    }
+    pub fn open(filename: Option<&[u8]>) -> Result<*mut u8, String> {
+        let prev_error_mode = unsafe {
+            // disable "dll load failed" error dialog.
+            // Note that SetErrorMode is process-wide so this can cause race condition!
+            // however, since even Windows APIs do not care of such problem (#20650),
+            // we just assume SetErrorMode race is not a great deal.
+            // (Windows 7 provides SetThreadErrorMode so we may use it if possible).
+            //
+            // SEM_FAILCRITICALERRORS 0x01
+            SetErrorMode(1)
+        };
+
+        unsafe {
+            SetLastError(0);
+        }
+
+        let result = match filename {
+            Some(filename) => {
+                let filename_str = str::from_utf8(filename).unwrap();
+                let mut filename_str: Vec<u16> = filename_str.utf16_units().collect();
+                filename_str.push(0);
+                let result = unsafe {
+                    LoadLibraryW(filename_str.as_ptr() as *const libc::c_void)
+                };
+                // beware: Vec/String may change errno during drop!
+                // so we get error here.
+                if result == ptr::null_mut() {
+                    Err(format!("Error code {}", os::errno()))
```

Comment:
While you're at it, could this get a beefed up error message with the error description as well as the error code?